python/dgl/convert.py

```diff
@@ -429,94 +244,236 @@ def hetero_from_shared_memory(name):
+
     if num_nodes_dict is None:
-        num_nodes_dict = defaultdict(int)
-        for (srctype, etype, dsttype), data in data_dict.items():
-            _, _, nsrc, ndst = data
-            num_nodes_dict[srctype] = max(num_nodes_dict[srctype], nsrc)
-            num_nodes_dict[dsttype] = max(num_nodes_dict[dsttype], ndst)
+        num_nodes_dict = num_nodes_dict_
+    else:
+        utils.check_type(num_nodes_dict, dict, 'num_nodes_dict', skip_none=False)
+        for nty in num_nodes_dict_:
+            if nty not in num_nodes_dict:
+                raise DGLError('Missing node type {} for num_nodes_dict'.format(nty))
+            if num_nodes_dict[nty] < num_nodes_dict_[nty]:
+                raise DGLError('Expect the number of nodes to be at least {:d} for type {}, '
+                               'got {:d}'.format(num_nodes_dict_[nty], nty, num_nodes_dict[nty]))
+
+    # Graph creation
+    # TODO(BarclayII): I'm keeping the node type names sorted because even if
+    # the metagraph is the same, the same node type name in different graphs may
+    # map to different node type IDs.
+    # In the future, we need to lower the type names into C++.
+    ntypes = list(sorted(num_nodes_dict.keys()))
+    num_nodes_per_type = utils.toindex([num_nodes_dict[ntype] for ntype in ntypes], "int64")
+    ntype_dict = {ntype: i for i, ntype in enumerate(ntypes)}
 
+    meta_edges_src = []
+    meta_edges_dst = []
+    etypes = []
     rel_graphs = []
     for (srctype, etype, dsttype), data in data_dict.items():
-        u, v, _, _ = data
-        if srctype == dsttype:
-            rel_graphs.append(graph(
-                (u, v), srctype, etype,
-                num_nodes=num_nodes_dict[srctype],
-                validate=validate,
-                formats=formats,
-                idtype=idtype, device=device))
-        else:
-            rel_graphs.append(bipartite(
-                (u, v), srctype, etype, dsttype,
-                num_nodes=(num_nodes_dict[srctype], num_nodes_dict[dsttype]),
-                validate=validate,
-                formats=formats,
-                idtype=idtype, device=device))
+        meta_edges_src.append(ntype_dict[srctype])
+        meta_edges_dst.append(ntype_dict[dsttype])
+        etypes.append(etype)
+        src, dst, _, _ = data
+        g = create_from_edges(src, dst, srctype, etype, dsttype,
+                              num_nodes_dict[srctype], num_nodes_dict[dsttype])
+        rel_graphs.append(g)
 
-    return hetero_from_relations(rel_graphs, num_nodes_dict)
+    # metagraph is DGLGraph, currently still using int64 as index dtype
+    metagraph = graph_index.from_coo(len(ntypes), meta_edges_src, meta_edges_dst, True)
+    # create graph index
+    hgidx = heterograph_index.create_heterograph_from_relations(
+        metagraph, [rgrh._graph for rgrh in rel_graphs], num_nodes_per_type)
+    retg = DGLHeteroGraph(hgidx, ntypes, etypes)
```

Comment:
Did you try creating a graph with edges already are gpu tensors?
Will this cause problem as hgidx here is in cpu, and will be moved to gpu later.

Reply:
I tried and it works.